### PR TITLE
Handle partial resolve cases

### LINKED ISSUE #2443: infer break use-as alias after macro expansion
```rust
#![allow(non_camel_case_types)]

mod m {
    macro_rules! _foo {
        ($x:ident) => { type $x = u64; }
    }
    pub(crate) use _foo as foo;
}

m::foo!(foo);
use foo as bar;

fn f() -> bar { 0 }
fn b() -> foo { 0 }

fn main() {  
    let _a  = f();      // _a : {unknown}
    let _b  = b();     // _b: u64
}
```

It is because `use foo as bar` is resolved before expansion of  `m::foo!(foo)`, and we didn't rewrite the aliased `PerNs` of `bar`. Note that legacy macro invocation (i.e. unquatified identifier like `foo!`) is not affected, because it is expanded before any item resolutions.

### PR DESCRIPTION
**Pull request by edwin0cheng**
Another try to fix #2443 :

We resolve all imports every time in `DefCollector::collect` loop even it is resolved previously.  
This is because other unresolved imports and macros will bring in another `PerNs`, so we can only assume that it has been partially resolved.

### CODE REVIEW

**Comment by matklad** [comment]
I wonder how this looks performance-wise. My gut feeling is that we'll need something more involved here, like tracking unresovled types/values/macros separatelly, as well as remembering name resolution results for prefixes (instead of resolving each path from the start every time).

**Comment by matklad** [comment]
Hm, another easy optimization we can do is to mark paths which are rooted at external crates as `ReachedFixedPoint::Yes` -- crate graphs can't change in other crates.

**Comment by edwin0cheng** [comment]
> Hm, another easy optimization we can do is to mark paths which are rooted at external crates as ReachedFixedPoint::Yes -- crate graphs can't change in other crates. 

Good idea. I added some easy optimizations:

* Skip adding it to unresolved if it is with external crates or external crates root
* Skip record import if the PerNs did not changed

> like tracking unresovled types/values/macros separatelly, as well as remembering name resolution results for prefixes (instead of resolving each path from the start every time).

I understand the name resolution prefixes one (I would recommend adding general path cache in DefCollector in another PR), but why should we tracks unresolved types/values/macros separately ?

**Comment by matklad** [comment]
Hm, I feel uneasy about this change, for two reasons:

* it adds even more elements to already large and unreadable tuples. This bit of code needs a redesign, it's already (before this PR) way to complex. 
* with this change, the unified `unresolved_imports` list doesn't actually make sense, as everything happens on per-ns basis. 

It might be a good idea to check, how this work list is setup in `rustc`. What exactly is the state during iteration?

**Comment by edwin0cheng** [comment]
> Hm, I feel uneasy about this change, for two reasons:

Well, these changes can of course be rejected because I don't have a general idea how to solve this problem and trying out to see what code will work.  :)

> It might be a good idea to check, how this work list is setup in rustc. What exactly is the state during iteration?

Sure, and you are right. `rustc` solve this problem very easily.

**TLDR:**

 here are an over simplifer version of how rustc do (pseudo code):
```rust
fn resolve_imports() {
    while old_unresolved_imports.len() != unresolved_imports.len() {
        for item in old_unresolved_imports {
            let res = resolve_item(item);
            if res == RESOLVED {
                unresolved_imports.push(item);
            } else {
                resolved_imports.push(item)
            }
        }
    };
}

fn passes() {
    // Expand macro pass
    loop {
        resolve_imports();
        
        if resolve_and_expand_macros() == REACH_FIX_POINT {
            break;
        }
    }
    
    unresolved_imports.extends(resolved_imports);
    resolve_imports();
}
```

**Details**

An import item is added in [add_import_directive](https://github.com/rust-lang/rust/blob/135ccbaca86ed4b9c0efaf0cd31442eae57ffad7/src/librustc_resolve/build_reduced_graph.rs#L307), the namespace is stored in [`ImportDirect::subclass`](https://github.com/rust-lang/rust/blob/135ccbaca86ed4b9c0efaf0cd31442eae57ffad7/src/librustc_resolve/build_reduced_graph.rs#L323), which is [empty](https://github.com/rust-lang/rust/blob/135ccbaca86ed4b9c0efaf0cd31442eae57ffad7/src/librustc_resolve/build_reduced_graph.rs#L471) in the beginning. 

While the expanding loop start at [`fully_expand_fragment`](https://github.com/rust-lang/rust/blob/f577b0ef6e637ab7a6095cdfe0b51fa3faf97050/src/libsyntax_expand/expand.rs#L362), it will call `resolve_imports` and try to expand macro invocations. [`resolve_import`](https://github.com/rust-lang/rust/blob/f577b0ef6e637ab7a6095cdfe0b51fa3faf97050/src/libsyntax_expand/expand.rs#L488) just forward to [`ImportResolver::resolver_import`](https://github.com/rust-lang/rust/blob/1e2a73867d9c0732c89da8de19d4f18a6e80dcfe/src/librustc_resolve/resolve_imports.rs#L623), and you can see it is a fixed-point algorithm. 

After all expansions loop ends, for each crate, `resolver.resolve_crate(krate)` will be [called](https://github.com/rust-lang/rust/blob/7afe6d9d1f48b998cc88fe6f01ba0082788ba4b9/src/librustc_interface/passes.rs#L425), and it will call [`finalize_imports`](https://github.com/rust-lang/rust/blob/809e180a76ce97340bf4354ff357bc59e3ca40b2/src/librustc_resolve/resolve_imports.rs#L636), basically what it do is resolve all imports again and do some checking and error reporting.

**Comment by matklad** [comment]
Thanks for this overview @edwin0cheng , it's really helpful!

So I think this are the main diff with our current mode:

1. Imports track fixed-point state per-ns, `indeterminate = true` is executed if at least one ns could be further resolved
2. Imports are tree-shaped. `ImportDirective`, I think, refers to a single *segment* of an import. This allows one to avoid repeatedly resolve the same path (which is something we do a lot)
3. An import is marked as `Err(Determined)`  if we know it can't resolve eo anything

Notably, it still maintains a single list of unresolved imports!

From the points above, `1` affects correctness, while `2` and `3` are strictly performance optimizations. 

I think we probably should start here with a solid foundation for `1`. 

`PartialResolvedImport` is sort-of goes in that direction, but I think it complicates the existing code, *without* solving the core problem (the fact, that we need to track three fixed-points per import).


I am out of steam for today to be able to concretely describe the next steps here, but I feel a set of pre-requsites is:

* just general refactorings, to make sure that we don't work with four-tuples everywhere
* some plumbing to make sure that we track `ReachedFixedPoint` per namespace
* initially, I think we still want to consider an imported resolved if at least *one* namespace is resolved (to avoid regressing perf)
* make sure that, when we resolve import in external crate, we mark all three nses as reached fixedpoint
* flipping the "resolved" condition for import from "at least one ns" to "all three ns"

**Comment by edwin0cheng** [comment]
I rebased and made some following changes:

1. Add `ImportDirective` for replace the  four-tuples
2. Added `PartialResolvedImport::Indeterminate` indicate it is not fully resolved.
3. Add check whether if all namespace all resolved, mark it as `PartialResolvedImport::Resolved`
4. We make the `resolve_imports` itself do the fixpoint algorithm. 
5. Although we mark PartialResolvedImport::Indeterminate is resolved. (By pushing it in `resolved_imports`, we add another resolve_imports after the resolve loop to resolve all indeterminate resolve  again to catch all missing cases. (That's why the new tests could pass)

**Comment by matklad** [comment]
This looks great now! 

bors r+

**Comment by edwin0cheng** [comment]
bors retry
